src/Http2Client.ts

```diff
@@ -40,38 +44,82 @@ export class Http2Client extends EventEmitter implements Client {
 	 *
 	 * @param host - hostname or IP address of daemon
 	 * @param port - port number of daemon on the host
-	 * @override
 	 */
-	introduce(host: string, port: number): void {
+	public introduce(host: string, port: number): void {
 		const client = connect(`http://${hostAndPort(host, port)}`)
 		client.on('error', (err) => this.emit('error', err))
-		this.available(client)
+		this.setAvailableAndCheckJobs(client)
 	}
 
 	/**
-	 * Drive the given daemon, by providing it with one task at a time in an async loop.
+	 * Mark the daemon as available.
+	 * Then calls function to check if there are any doable jobs.
 	 *
-	 * @param client - a HTTP/2 client connected to a daemon.
+	 * @param daemon - The daemon that is available.
+	 */
+	private setAvailableAndCheckJobs(daemon: ClientHttp2Session): void {
+		this.availableDaemons.add(daemon)
+		this.checkJobsAndAssign()
+	}
+
+	/**
+	 * Tries to assign the next available job to a daemon.
+	 * Closes all daemons if all jobs are finished.
 	 */
-	private available(client: ClientHttp2Session): void {
-		const job = this.queue.pop()
-		if (!job) {
-			client.close(() => this.emit('done'))
-			return
+	private checkJobsAndAssign(): void {
+		const job = this.jobOrderer.popNextJob()
+
+		if (job && this.availableDaemons.size > 0) {
```

Comment:
No because if no Jobs are currently available, none will become available until another Job completes. If a job completes then the Daemon must have responded to our request. Every time we get a response we run `setAvailableAndCheckJobs()`.

Reply:
I'm thinking of the case where one job finishes and unlocks multiple other jobs that were depending on it. In that case, the job finishing will call this function, but it will only pick up one of the potentially many jobs now available.

Reply:
Yeah you're totally right. I just pushed some code to fix it.